Updated the clustering file

sangeetaraghu opened the pull request:
@tsegismont As discussed, I have push the clsutering adoc file with some changes.
I have also changed the file name. Could you please update the index.adoc as well?

tsegismont commented on asciidoc/modules/con_mg_generic-updates-in-vertx-http.adoc:
```diff
@@ -109,11 +109,13 @@ The following example shows how to set WebSocket connections in {VertX} {v4}.
+`HttpMethod` is available as a new interface.
 
-If you are using `HttpMethod` enumerated data type in a switch block, you can use the following code to migrate your application to {VertX} {v4}.
+In earlier releases of {VertX}, `HttpMethod` was declared as an enumerated data type. As an enumeration, it limited the extensibility of HTTP. Further, it prevented serving other HTTP methods with this type directly. You had to use the  `HttpMethod.OTHER` value along with the `rawMethod` attribute during server and client HTTP requests.
```
Can you move the changes in this file to another PR?

sangeetaraghu replied:
@tsegismont This file does not require review. I rephrased some sections. Is it okay to merge it in this PR?

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+= Migrating clusters
+
+It is important to decide the migration strategy for your codebase in a cluster. This is because you cannot add {VertX} {v3x} nodes and {VertX} {v4} nodes together in a single cluster for the following reasons: 
```
```suggestion
It is important to decide the migration strategy for your codebase. This is because you cannot add {VertX} {v3x} nodes and {VertX} {v4} nodes together in a single cluster for the following reasons:
```

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+It is important to decide the migration strategy for your codebase in a cluster. This is because you cannot add {VertX} {v3x} nodes and {VertX} {v4} nodes together in a single cluster for the following reasons: 
+
+* Cluster manager upgrades - If there are major version upgrades in cluster managers, they may not have support all features for backward compatibility. You will have to check the updates in cluster managers.
```
```suggestion
* Cluster manager upgrades - major version upgrades in cluster managers prevent with non backward-compatible changes
```

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+It is important to decide the migration strategy for your codebase in a cluster. This is because you cannot add {VertX} {v3x} nodes and {VertX} {v4} nodes together in a single cluster for the following reasons: 
+
+* Cluster manager upgrades - If there are major version upgrades in cluster managers, they may not have support all features for backward compatibility. You will have to check the updates in cluster managers.
+* Subscription data changes - If the information stored in subscription maps or caches is no longer the same in the cluster manager versions. 
```
```suggestion
* Subscription data changes - {VertX} has changed the format of the EventBus subscription data stored in cluster managers. 
```

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+* Cluster manager upgrades - If there are major version upgrades in cluster managers, they may not have support all features for backward compatibility. You will have to check the updates in cluster managers.
+* Subscription data changes - If the information stored in subscription maps or caches is no longer the same in the cluster manager versions. 
+* Transport protocol changes - If some fields in the message transport protocol have changed in the cluster.
```
```suggestion
* Transport protocol changes - {VertX} has changed some fields in the message transport protocol in the cluster.
```

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+== Splitting the cluster
+
+If you have a cluster where different teams have deployed verticles for their applications, you can consider splitting the {VertX} {v3x} cluster into smaller ones. Note that after splitting the cluster, the separated components will not be able to communicate using the clustering features. You can splitthe cluster using the following XXX:
```
```suggestion
If you have a cluster where different teams have deployed verticles for their applications, you can consider splitting the {VertX} {v3x} cluster into smaller ones. Note that after splitting the cluster, the separated components will not be able to communicate using the clustering features. You can split the cluster using the following XXX:
```

tsegismont replied:
What does XXX stand for?

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+If you have a cluster where different teams have deployed verticles for their applications, you can consider splitting the {VertX} {v3x} cluster into smaller ones. Note that after splitting the cluster, the separated components will not be able to communicate using the clustering features. You can splitthe cluster using the following XXX:
+
+* EventBus request and reply - HTTP/RESTful web services, gRPC
+* EventBus send and publish - Messaging systems such as, Kafka, Postgres `LISTEN` and `NOTIFY`, Redis Pub and Sub
```
```suggestion
* EventBus send and publish - Messaging systems such as Apache Kafka or Apache ActiveMQ, Postgres `LISTEN` and `NOTIFY`, Redis Pub and Sub
```

sangeetaraghu replied:
As discussed, adding an attribute to show "Apache Kafka or Apache ActiveMQ" only in community documentation.

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+* EventBus request and reply - HTTP/RESTful web services, gRPC
+* EventBus send and publish - Messaging systems such as, Kafka, Postgres `LISTEN` and `NOTIFY`, Redis Pub and Sub
+* Shared Data - Redis, Infinispan
```
```suggestion
* Shared Data - Redis, Infinispan, Hazelcast
```

sangeetaraghu replied:
As discussed, adding an attribute to show "Hazelcast" only in community documentation.

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+* EventBus send and publish - Messaging systems such as, Kafka, Postgres `LISTEN` and `NOTIFY`, Redis Pub and Sub
+* Shared Data - Redis, Infinispan
+
+When you split the cluster, each team can move to {VertX} {v4} when they are ready or if it is needed.
```
```suggestion
After you split the cluster, each team can move to {VertX} {v4} when they are ready or if it is needed.
```

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+== Using {VertX} EventBus Link
+
+If you cannot split your cluster, then use https://github.com/vert-x3/vertx-eventbus-link[Vert.x EventBus Link] to migrate your codebase incrementallyXXX.
```
XXX ?

sangeetaraghu replied:
@tsegismont to migrate your codebase **incrementally**
This rephrase is ok?- -> incrementally?

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+WARNING: The migration of shared data API, that is, maps, counters and locks is not supported.
+
+The tool creates an `EventBusLink` object that implements the `EventBus` interface. An instance of `EventBusLink` is created on one node of each cluster. The instance is created by providing a set of addresses and its behavior depends on the message paradigmXXX:
```
```suggestion
The tool creates an `EventBusLink` object that implements the `EventBus` interface. An instance of `EventBusLink` is created on at least one node of each cluster. The instance is created by providing a set of addresses and its behavior depends on the message paradigmXXX:
```

tsegismont replied:
XXX ?

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+The tool creates an `EventBusLink` object that implements the `EventBus` interface. An instance of `EventBusLink` is created on one node of each cluster. The instance is created by providing a set of addresses and its behavior depends on the message paradigmXXX:
+
+* _fire and forget_ and _request/reply_ - The message is sent (XXX instaed of forwarded) to the remote cluster which delegates (XXX what?) to the {VertX} EventBus.
```
```suggestion
* _fire and forget_ and _request/reply_ - The message is sent to the remote cluster
```

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+The tool creates an `EventBusLink` object that implements the `EventBus` interface. An instance of `EventBusLink` is created on one node of each cluster. The instance is created by providing a set of addresses and its behavior depends on the message paradigmXXX:
+
+* _fire and forget_ and _request/reply_ - The message is sent (XXX instaed of forwarded) to the remote cluster which delegates (XXX what?) to the {VertX} EventBus.
+* _publish_ - The message is sent (XXX instaed of forwarded) to the remote cluster and to the {VertX} EventBus.
```
```suggestion
* _publish_ - The message is sent to both this cluster and the remote cluster.
```

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+The {VertX} EventBus Link creates a WebSocket server to receive messages and uses a WebSocket client to send them.
+
+See the sections link:https://github.com/vert-x3/vertx-eventbus-linkdocumentation for more details.
```
There are references to the different sections missing, I think

sangeetaraghu replied:
Updated the links.

tsegismont reviewed (changes requested):
Thank you @sangeetaraghu . I made some suggestions

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+It is important to decide the migration strategy for your codebase. This is because you cannot add {VertX} {v3x} nodes and {VertX} {v4} nodes together in a single cluster for the following reasons:
+
+* Cluster manager upgrades - Major version upgrades in cluster managers prevent breaking (non-backward compatibility) changes.
```
```suggestion
* Cluster manager upgrades - Major version upgrades in cluster managers prevent backward compatibility.
```

tsegismont commented on asciidoc/modules/con_mg_migrating-clusters.adoc:
```diff
@@ -0,0 +1,40 @@
+
+* EventBus request and reply - HTTP/RESTful web services, gRPC
+* EventBus send and publish - Messaging systems such as, {AMQ}, Postgres `LISTEN` and `NOTIFY`, Redis Pub and Sub
+* Shared Data - Redis, Infinispan, {Hazelcast}
```
Shouldn't the comma and following space be part of the `Hazelcast` variable? Otherwise if the variable is not set there will be a trailing comma

vietj replied:
just remove the implementation names and avoid using variables for that.

vietj replied:
meaning that having a variable just to avoid hazelcast or redis looks awkward

tsegismont reviewed (changes requested):
A couple more comments @sangeetaraghu

sangeetaraghu commented:
@vietj @tsegismont I have updated the section and removed the attributes. Can you please review the PR? If you agree to the final draft we can close the PR.

tsegismont reviewed (approved):
Thank you @sangeetaraghu

tsegismont commented:
@sangeetaraghu before merging, you must:

- update `assembly_mg_changes-in-clustering.adoc` to resolve the conflict
- remove the `con_mg_migrating-clusters-gradually.adoc` file
